### PR TITLE
las: improve coverage

### PR DESCRIPTION
**Pull request by wilzbach**
Should increase the coverage by quite a bit.

### CODE REVIEW

File: source/mir/las/sum.d [hunk 1]
```diff
@@ -1762,8 +1762,8 @@ public:
 nothrow unittest
 {
     import std.algorithm.iteration: map;
-    import std.range;
-    import std.math;
```
Comment (wilzbach) [inline source/mir/las/sum.d:1766]: not needed.

File: source/mir/las/sum.d [hunk 2]
```diff
@@ -1782,18 +1782,56 @@ nothrow unittest
...
+            assert(summator.isNaN() == r.isNaN());
+            assert(summator.isFinite() == r.isFinite());
+            assert(summator.isInfinity() == r.isInfinity());
+            assert(s == r || s.isNaN && r.isNaN);
+        }
+        summator = 0.0;
+    }
+}
+
+nothrow unittest
```
Comment (wilzbach) [inline source/mir/las/sum.d]: I splitted the tests infinity tests out and run them for the pairwise summator too. As we can't just create an array of them, `runTests` was created.

File: source/mir/las/sum.d [hunk 3]
```diff
@@ -1803,16 +1841,55 @@ nothrow unittest
...
+    foreach (i, test; tests)
+    {
+        runTests(summatorPrecise, test);
+        runTests(summatorPairwise, test);
+    }
+}
+
+nothrow unittest
+{
+    // different result than precise
```
Comment (wilzbach) [inline source/mir/las/sum.d]: I wasn't sure whether this is a bug or intended behavior - please check.

File: source/mir/las/sum.d [hunk 4]
```diff
@@ -1988,6 +2065,28 @@ unittest
         assert(s3.array == c.array);
     }
 }
+unittest
+{
+    import core.simd;
+    import std.range: iota;
+    import std.array: array;
+
+    double[] ns = [9.0, 101.0];
```
Comment (wilzbach) [inline source/mir/las/sum.d:2061]: there are quite many specializations for `pairwise` - here we add tests for the non-tested parts.

File: source/mir/las/sum.d [hunk 5]
```diff
@@ -1782,18 +1782,56 @@ nothrow unittest
         tuple([2.0^^53+10.0, 1.0, 2.0^^-100], 2.0^^53+12.0),
         tuple([2.0^^53-4.0, 0.5, 2.0^^-54], 2.0^^53-3.0),
         tuple([M-2.0^^970, -1, M], 1.7976931348623157e+308),
-        tuple([double.max, double.max*2.^^-54], double.max),
-        tuple([double.max, double.max*2.^^-53], double.infinity),
```
Comment (9il) [inline source/mir/las/sum.d:1786]: Please don,t cut tests from here. Just copy-past them for Pairwise summation

Comment (9il) [inline source/mir/las/sum.d:1786]: BTW, don't use this test for Pairwise summation  because they may gail on the future

Comment (9il) [inline source/mir/las/sum.d:1786]: results depends on grouping order.

**Comment by 9il** [comment]
lets postpone ndslice fixes until Phobos PR. just for forward with `las.sum`

File: source/mir/las/sum.d [hunk 6]
```diff
@@ -1815,6 +1815,98 @@ nothrow unittest
...
+            assert(summator.isInfinity() == r.isInfinity());
+            assert(s == r || s.isNaN && r.isNaN);
+        }
+        summator = 0.0;
+    }
+}
+
+nothrow unittest
+{
+    // unittest with different result to precise summation
```
Comment (wilzbach) [inline source/mir/las/sum.d]: using kahan summation quite many tests are different...

File: source/mir/las/sum.d (hunk 5 above)
Comment (wilzbach) [inline source/mir/las/sum.d:1786]: > Please don,t cut tests from here. Just copy-past them for Pairwise summation

Okay - just thought this is bad ;-)

**Comment by wilzbach** [comment]
> lets postpone ndslice fixes until Phobos PR. just for forward with las.sum

Sorry for the bad timing.

File: source/mir/las/sum.d [hunk 7]
```diff
@@ -1815,6 +1815,98 @@ nothrow unittest
...
+        tuple([M-2.0^^970, 1.0, M], double.infinity),
+        tuple([M, M], double.infinity),
+        tuple([-M, 2.^^971, -M], -double.max),
+    ];
+    foreach (i, test; tests)
+    {
+        foreach (t; test[0]) summator.put(t);
+        auto r = test[1];
+        auto s = summator.sum;
+        version(X86)
```
Comment (9il) [inline source/mir/las/sum.d]: remove version `version` constraints here

File: source/mir/las/sum.d [hunk 8]
```diff
@@ -1815,6 +1815,98 @@ nothrow unittest
...
+    Summator!(double, Summation.kahan) summator = 0.0;
+
+    enum double M = (cast(double)2) ^^ (double.max_exp - 1);
+    Tuple!(double[], double)[] tests = [
+        tuple(new double[0], 0.0),
+        tuple([0.0], 0.0),
+        tuple([double.max, double.max*2.^^-54], double.max),
+        tuple([double.max, double.max*2.^^-53], double.infinity),
+        tuple(iota(1, 1001).map!(a => 1.0/a).array , 7.4854708605503451),
+        tuple(iota(1, 1001).map!(a => (-1.0)^^a/a).array, -0.69264743055982025), //0.693147180559945309417232121458176568075500134360255254120680...
```
Comment (9il) [inline source/mir/las/sum.d]: remove all tests with iota from here

File: source/mir/las/sum.d (hunk 7 above)
Comment (wilzbach) [inline source/mir/las/sum.d]: (this was copied from the test above)

File: source/mir/las/sum.d [hunk 9]
```diff
@@ -1815,6 +1815,98 @@ nothrow unittest
...
+        version(X86)
+        {
+            assert(summator.isNaN() == r.isNaN());
+            assert(summator.isFinite() == r.isFinite() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(summator.isInfinity() == r.isInfinity() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(nextDown(s) <= r && r <= nextUp(s) || s.isNaN && r.isNaN);
+        }
+        else
+        {
+            assert(summator.isNaN() == r.isNaN());
```
Comment (9il) [inline source/mir/las/sum.d]: just use else branch

File: source/mir/las/sum.d (hunk 7 above)
Comment (9il) [inline source/mir/las/sum.d]: i know

File: source/mir/las/sum.d [hunk 10]
```diff
@@ -1815,6 +1815,98 @@ nothrow unittest
...
+        auto r = test[1];
+        auto s = summator.sum;
+        version(X86)
+        {
+            assert(summator.isNaN() == r.isNaN());
+            assert(summator.isFinite() == r.isFinite() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(summator.isInfinity() == r.isInfinity() || r == -double.max && s == -double.infinity || r == double.max && s == double.infinity);
+            assert(nextDown(s) <= r && r <= nextUp(s) || s.isNaN && r.isNaN);
+        }
+        else
```
Comment (9il) [inline source/mir/las/sum.d]: just use else branch (always for kahan, KBN, KB2)

File: source/mir/las/sum.d (hunk 6 above)
Comment (9il) [inline source/mir/las/sum.d]: Unite this test for kahan with previous one and generalize it with all except `precise`

File: source/mir/las/sum.d [hunk 11]
```diff
@@ -1832,10 +1833,74 @@ nothrow unittest
...
+
+    }
+    foreach (i, test; tests)
+    {
+        runTest(summatorKahan, test);
+        runTest(summatorPairwise, test);
+        runTest(summatorNaive, test);
+        runTest(summatorFast, test);
+        //runTest(summatorKbn, test);
+        //runTest(summatorKb2, test);
```
Comment (wilzbach) [inline source/mir/las/sum.d]: Those two algorithms have a lot of different results.

File: source/mir/las/sum.d [hunk 12]
```diff
@@ -1832,10 +1833,74 @@ nothrow unittest
...
+        //tuple([M, M, M, M, -M, -M], double.nan),
+        //tuple([M, M, M, M, -M, M], double.nan),
+        //tuple([-M, -M, -M, -M], -double.nan),
+        //tuple([M, M, -2.^^971], double.nan),
+        //tuple([M, M, -2.^^970], double.nan),
+        //tuple([-2.^^970, M, M, -2.^^-1074], double.nan),
+        //tuple([M, M, -2.^^970, 2.^^-1074], double.nan),
+        //tuple([-M, -M, 2.^^970], -double.nan),
+        //tuple([-M, -M, 2.^^970, 2.^^-1074], -double.nan),
+        //tuple([-2.^^-1074, -M, -M, 2.^^970], -double.nan),
```
Comment (wilzbach) [inline source/mir/las/sum.d:1874]: somehow this part is different depending on the algorithm..

File: source/mir/las/sum.d [hunk 13]
```diff
@@ -1988,6 +2053,28 @@ unittest
...
+    double[] ns = [9.0, 101.0];
+
+    foreach (n; ns)
+    {
+        double[] ar = iota(n).array;
+        double c = n * (n - 1) / 2; // gauss for n=100
+        double s0 = ar.sum!(Summation.pairwise);
+        double s1 = ar.sum!(Summation.kahan);
+        double s2 = ar.sum!(Summation.naive);
+        double s3 = ar.sum!(Summation.fast);
```
Comment (9il) [inline source/mir/las/sum.d]: KBN and KB2?

Comment (9il) [inline source/mir/las/sum.d]: and precise

Comment (9il) [inline source/mir/las/sum.d]: BTW, please use `AliasSeq` to reduce code size

Comment (wilzbach) [inline source/mir/las/sum.d]: > KBN and KB2?

https://github.com/DlangSciene/mir/blob/improve_cov4/source/mir/las/sum.d#L324
https://github.com/DlangSciene/mir/blob/improve_cov4/source/mir/las/sum.d#L352

Added precise - can I remove the redundant parts from the precise tests?

Comment (9il) [inline source/mir/las/sum.d]: > Added precise - can I remove the redundant parts from the precise tests?

I am about test with `double[] ns = [9.0, 101.0];`

Comment (wilzbach) [inline source/mir/las/sum.d]: > BTW, please use AliasSeq to reduce code size

I was looking for that - thanks :)

Comment (wilzbach) [inline source/mir/las/sum.d]: > KBN and KB2?

sorry - somehow github displayed this at the wrong point. now it should be there ;-)

**Comment by 9il** [comment]
Please rebase

**Comment by wilzbach** [comment]
> Please rebase

Oh I thought we first fix the problem in the infty test. Rebased.

**Comment by 9il** [comment]
> This branch is out-of-date with the base branch

have not rebased from master

> problem in the infty test

what problem?

**Comment by wilzbach** [comment]
> have not rebased from master

Sorry had a short issue with my git client - should be changed now.

> what problem?

I commented out some tests as they have different result for different summation algorithms.

**Comment by 9il** [comment]
> I commented out some tests as they have different result for different summation algorithms.

this is OK

**Comment by wilzbach** [comment]
> this is OK

(:

Just out of interest:  "0 of 2 checks passed" - I thought github would prevent you from merging then?

**Comment by 9il** [comment]
> Just out of interest: "0 of 2 checks passed" - I thought github would prevent you from merging then?

I don't know)))

**Pull request merged**